### PR TITLE
Support schema inheritance

### PR DESCRIPTION
**Pull request by nati**
We support schema inheritance for helping schema
development.

Closes: #7

### CODE REVIEW

**Comment by famao3** [comment]
does gohan support only go 1.5 or above ?
I tried compiling with go 1.4.2 but errors happened.

**Comment by nati** [comment]
Could you share the error? 
I've used go 1.4.2.

**Comment by famao3** [comment]
merging feature/issue7 to current master branch .

~~~
commit 647c99269632aa6b61eaa98b11d200ba031f2b19
Merge: 4fd3531 6af6f60
Author: root <root@gohan-heat-test.novalocal>
Date:   Fri Jan 22 18:38:33 2016 +0000

    Merge remote-tracking branch 'nati/feature/issue7'
~~~

make

~~~
root@gohan-heat-test:/go/src/github.com/cloudwan/gohan# make
==> Formatting
go fmt ./...
==> Linting
golint ./...
go vet ./...
==> Installing dependencies
godep restore
==> Building
go build ./...
# github.com/cloudwan/gohan/gohan/extension/otto
gohan/extension/otto/gohan_core.go:46: schema.RawData undefined (type *schema.Schema has no field or method RawData)
gohan/extension/otto/gohan_util.go:79: schema.RawData undefined (type *schema.Schema has no field or method RawData)
# github.com/cloudwan/gohan/gohan/server
gohan/server/editor.go:37: undefined: schema.GetSchema
gohan/server/editor.go:56: undefined: schema.GetSchema
make: *** [build] Error 2
~~~

**Comment by nati** [comment]
Could you try to delete your repo & clone again?

File: util/util.go [hunk 1]
```diff
@@ -227,3 +227,81 @@ func Match(a interface{}, b interface{}) bool {
...
+		}
+	}
+	return false
+}
+
+//ExtendStringList appends a value if it isn't in base list
+func ExtendStringList(target []string, list []string) []string {
+	extended := make([]string, 0, len(target))
+	if target != nil {
+		for _, item := range target {
```
Comment (pedro-r-marques) [inline util/util.go]: May using builtin.copy() would be preferable ?

File: util/util.go [hunk 2]
```diff
@@ -227,3 +227,81 @@ func Match(a interface{}, b interface{}) bool {
...
+			return true
+		}
+	}
+	return false
+}
+
+//ExtendStringList appends a value if it isn't in base list
+func ExtendStringList(target []string, list []string) []string {
+	extended := make([]string, 0, len(target))
+	if target != nil {
```
Comment (pedro-r-marques) [inline util/util.go]: Does it make sense for target to be nil ?

File: util/util.go [hunk 3]
```diff
@@ -227,3 +227,81 @@ func Match(a interface{}, b interface{}) bool {
...
+				extended = append(extended, item)
+			}
+		}
+	}
+
+	return extended
+}
+
+//ExtendMap extends a map from existing one
+func ExtendMap(target map[string]interface{}, object map[string]interface{}) map[string]interface{} {
```
Comment (pedro-r-marques) [inline util/util.go]: would it make sense to declare the arguments as (target, object map[string]interface{}) ?

File: docs/source/schema.rst [hunk 4]
```diff
@@ -178,6 +180,120 @@ schema prefix- see :ref:`namespace section <section-namespace>` for details.
 
 You can use following properties in json schema.
 
+Schema Inheritance
+-------------------------------
+
+You can make abstract schema by set type="abstract".
```
Comment (pedro-r-marques) [inline docs/source/schema.rst]: can define an abstract schema by setting...

File: docs/source/schema.rst [hunk 5]
```diff
@@ -178,6 +180,120 @@ schema prefix- see :ref:`namespace section <section-namespace>` for details.
...
+You can construct new schema from abstract schema using "extends".
```
Comment (pedro-r-marques) [inline docs/source/schema.rst]: Schemas can be derived from an abstract schema by using the keyword "extends".

File: docs/source/schema.rst [hunk 6]
```diff
@@ -178,6 +180,120 @@ schema prefix- see :ref:`namespace section <section-namespace>` for details.
...
+JSON schema, Metadata ,and Action will be merged when you extends schema.
```
Comment (pedro-r-marques) [inline docs/source/schema.rst]: JSON schema, metadata and action will be merged when a schema is extended.

File: docs/source/schema.rst [hunk 7]
```diff
@@ -178,6 +180,120 @@ schema prefix- see :ref:`namespace section <section-namespace>` for details.
...
+prefix value and parent value will be set if it isn't specifed in a schema.
```
Comment (pedro-r-marques) [inline docs/source/schema.rst]: if not specified

File: util/util.go (hunk 1 above)
Comment (nati) [inline util/util.go]: done

File: util/util.go (hunk 2 above)
Comment (nati) [inline util/util.go]: done

File: util/util.go (hunk 3 above)
Comment (nati) [inline util/util.go]: done

File: docs/source/schema.rst (hunk 4 above)
Comment (nati) [inline docs/source/schema.rst]: done

File: docs/source/schema.rst (hunk 5 above)
Comment (nati) [inline docs/source/schema.rst]: done

File: docs/source/schema.rst (hunk 6 above)
Comment (nati) [inline docs/source/schema.rst]: done

Comment (nati) [inline docs/source/schema.rst]: done

File: docs/source/schema.rst (hunk 7 above)
Comment (nati) [inline docs/source/schema.rst]: done

**Comment by pedro-r-marques** [comment]
LGTM

**Pull request merged**